interface/main/holidays/Holidays_Controller.php

```diff
@@ -69,33 +91,139 @@ public function upload_csv($files)
+    public function get_file_csv_data(): array
     {
         $file = [];
         if (file_exists($this->target_file)) {
-            $file['date'] = date("d/m/Y H:i:s", filemtime($this->target_file));
+            $file["date"] = date("d/m/Y H:i:s", filemtime($this->target_file));
         }
 
         return $file;
     }
 
+    public function get_last_error(): string
+    {
+        return $this->last_error;
+    }
+
+    private function is_valid_csv_upload(array $file): bool
+    {
+        if (!empty($file["error"])) {
+            $this->last_error = xl("Upload failed");
+            return false;
+        }
+
+        if (empty($file["tmp_name"]) || !is_uploaded_file($file["tmp_name"])) {
+            $this->last_error = xl("Invalid upload");
+            return false;
+        }
+
+        $name = $file["name"] ?? "";
+        $extension = strtolower(pathinfo($name, PATHINFO_EXTENSION));
+        if ($extension !== "csv") {
+            $this->last_error = xl("File must be a CSV");
+            return false;
+        }
+
+        return $this->is_valid_csv_content($file["tmp_name"]);
+    }
+
+    private function is_valid_csv_content($path): bool
+    {
+        $handle = fopen($path, "r");
+        if ($handle === false) {
+            $this->last_error = xl("Unable to read uploaded file");
+            return false;
+        }
+
+        try {
+            $row = null;
+            while (($data = fgetcsv($handle, 1000, ",", "'", "\\")) !== false) {
+                if (trim($data[0] ?? "") !== "") {
+                    $row = $data;
+                    break;
+                }
+            }
+
+            if ($row === null) {
+                $this->last_error = xl("CSV file is empty");
+                return false;
+            }
+
+            if ($this->is_header_row($row)) {
+                $row = null;
+                while (
+                    ($data = fgetcsv($handle, 1000, ",", "'", "\\")) !== false
+                ) {
+                    if (trim($data[0] ?? "") !== "") {
+                        $row = $data;
+                        break;
+                    }
+                }
+            }
+
+            if ($row === null || count($row) < 2) {
+                $this->last_error = xl("CSV row must have date and description");
+                return false;
+            }
+
+            $date = trim($row[0]);
+            if (!$this->is_valid_holiday_date($date)) {
+                $this->last_error = xl("Invalid date format in CSV");
+                return false;
+            }
+
+            return true;
+        } finally {
+            fclose($handle);
+        }
+    }
+
+    private function is_header_row(array $row): bool
+    {
+        $first = strtolower(trim($row[0] ?? ""));
+        $second = strtolower(trim($row[1] ?? ""));
+        return $first === "date" && $second === "description";
+    }
+
+    private function is_valid_holiday_date($date): bool
```

Comment:
```suggestion
    private function is_valid_holiday_date(string $date): bool
```

type declarations are necessary so that rector doesn't have to do casts and gives the right corrections.